be/src/exprs/function_call_expr.cpp

```diff
@@ -173,4 +176,64 @@ StatusOr<ColumnPtr> VectorizedFunctionCallExpr::evaluate_checked(starrocks::Expr
     return result;
 }
 
+bool VectorizedFunctionCallExpr::ngram_bloom_filter(ExprContext* context, const BloomFilter* bf,
+                                                    size_t index_gram_num) const {
+    FunctionContext* fn_ctx = context->fn_context(_fn_context_index);
+    std::vector<Slice>& ngram_set = fn_ctx->get_ngram_set();
+
+    const auto& gram_num_column = fn_ctx->get_constant_column(2);
+    // case like ngram_search(col,"needle", 5) when col has a 4gram bloom filter, don't use this index
+    if (gram_num_column != nullptr) {
+        size_t predicate_gram_num = ColumnHelper::get_const_value<TYPE_INT>(gram_num_column);
+        if (index_gram_num != predicate_gram_num) {
+            return true;
+        }
+    }
+
+    if (UNLIKELY(ngram_set.size() == 0)) {
+        Slice needle;
+        if (_fn_desc->name == "like" || _fn_desc->name == "regex") {
+            // not implemented right now
+            return true;
+        } else {
+            // checked in support_ngram_bloom_filter(size_t gram_num), so it 's safe to get const column's value
+            const auto& needle_column = fn_ctx->get_constant_column(1);
+            needle = ColumnHelper::get_const_value<TYPE_VARCHAR>(needle_column);
+        }
+
+        std::vector<size_t> index;
+        size_t slice_gram_num = get_utf8_index(needle, &index);
+        ngram_set.reserve(slice_gram_num - index_gram_num + 1);
+
+        size_t j;
+        for (j = 0; j + index_gram_num <= slice_gram_num; j++) {
+            // find next ngram
+            size_t cur_ngram_length = j + index_gram_num < slice_gram_num ? index[j + index_gram_num] - index[j]
+                                                                          : needle.get_size() - index[j];
+            Slice cur_ngram = Slice(needle.data + index[j], cur_ngram_length);
```

Comment:
it seems no need to save ngrams into ngram_set,   just invoke bf->test_bytes on cur_ngram is ok.